### PR TITLE
Add gc report

### PR DESCRIPTION
**Pull request by lynnro314**
closes part of #3127 
Added reports at the end of gc run, reporting the deleted objects, relevant commits and the gc rules. 
Missing commit timestamp in the commits summary.

### CODE REVIEW

**Comment by talSofer** [comment]
@lynnro314 can you please add an example report to the pr description? It will make it easier to review

**Review by Jonathan-Rosenberg — APPROVED** [review]
Cool 😺 !

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 1]
```diff
@@ -446,4 +463,30 @@ object GarbageCollector {
 
     bulkRemove(df, MaxBulkSize, bucket, region, awsRetries, snPrefix, hcValues).toDF("addresses")
   }
+
+  private def writeParquetReport(
+      dstRoot: String,
+      df: DataFrame,
+      time: String,
+      suffix: String = ""
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:471]: Nit: The location of `df` makes it a bit hard to read calls to the function.  I think I might prefer
```suggestion
  private def writeParquetReport(
      df: DataFrame,
      dstRoot: String,
      time: String,
      suffix: String = ""
```

Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:471]: (Also the same in `writeJsonSummary`, of course...).

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 2]
```diff
@@ -446,4 +463,30 @@ object GarbageCollector {
...
+  }
+
+  private def writeJsonSummary(
+      dstRoot: String,
+      numDeletedObjects: Long,
+      gcRules: String,
+      hcValues: Broadcast[ConfMap],
+      time: String
+  ) = {
+    val dstPath = new Path(dstRoot + s"/dt=${time}/summary.json")
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:484]:
```suggestion
    val dstPath = new Path(s"${dstRoot}/dt=${time}/summary.json")
```

Also, I really like interpolating values into strings like this -- can we do it in the paths above, please, for consistency?

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 3]
```diff
@@ -446,4 +463,30 @@ object GarbageCollector {
...
+    val dstFS = dstPath.getFileSystem(configurationFromValues(hcValues))
+    val jsonSummary = JObject("gc_rules" -> gcRules, "num_deleted_objects" -> numDeletedObjects)
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:486]: This is a bit strange: we just got `gcRules`, a value that is explicitly JSONable.  We transformed it into a string and then include this stringified value inside our real JSON.  The problem is that JSON-in-JSON is pretty awful to read (every quote is backslashed, newlines added for formatting before less readable, etc.).  Can we instead just stuff gcRules in some JSONified form here?

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 4]
```diff
@@ -446,4 +463,30 @@ object GarbageCollector {
...
+
+    val stream = dstFS.create(dstPath)
+    stream.writeChars(compact(render(jsonSummary)))
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:489]: This leaks `stream` if it throws an exception.  It should catch exceptions and abort the stream on failure.

**Review by arielshaqed — APPROVED** [review]
Thanks!

**Comment by lynnro314** [comment]
> @lynnro314 can you please add an example report to the pr description? It will make it easier to review

Sure.
1. Parquet files describing all the deleted objects:
image

2. A list of all the commits participating in this GC job:
image

3. Basic information about the run, including: # of deleted objects, the GC rules used for this jobs:
image
The rules:
image

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 5]
```diff
@@ -353,13 +361,21 @@ object GarbageCollector {
...
+    writeParquetReport(expiredAddresses, reportExpiredDst, time)
+    writeJsonSummary(reportLogsDst, removed.count(), gcRules, hcValues, time)
+
     removed
       .withColumn("run_id", lit(runID))
       .write
       .partitionBy("run_id")
       .mode(SaveMode.Overwrite)
-      .parquet(gcAddressesLocation + ".deleted")
+      .parquet(s"${storageNamespace}/_lakefs/logs/gc/deleted_objects/${time}/deleted.parquet")
```
Comment (talSofer) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:378]: Can you reuse the writeParquetReport to writing the deleted objects report? you can adjust it to receive optional partitionedBy param etc

**Review by talSofer — APPROVED** [review]
LGTM, thanks! 
Added one comment that you may want to consider addressing

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 5 above)
Comment (lynnro314) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:378]: I thought about that, but didn't find how I can do just `df.write.parquet(dstPath)` in that case..

**Pull request merged**